### PR TITLE
Atom: revamp services and add *Compatible types.

### PR DESCRIPTION
**Pull request by GlenCFL**
- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes:
[Point-compatible Types](https://github.com/atom/text-buffer/blob/7e72dc5e42ee53141f36dcc8f7cfb7d5aaa1f678/src/point.coffee#L3)
[Range-compatible Types](https://github.com/atom/text-buffer/blob/7e72dc5e42ee53141f36dcc8f7cfb7d5aaa1f678/src/range.coffee#L6)
[Linter Atom Package](https://atom.io/packages/linter)
[Autocomplete Plus Atom Package](https://atom.io/packages/autocomplete-plus)
[Status Bar Atom Package](https://atom.io/packages/status-bar)
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

This revamps the way Atom is going to provide definitions for other Atom packages (which provide services), with the ES6 import style now being used. I was previously going to use module augmentation on the global Atom namespace instead, but this seems like a better idea.

RangeCompatible and PointCompatible were both added and a lot of refactoring was done across AtomCore and TextBuffer as a result.

Finally, with the removal of editorconfigs in a previous PR (#20380), all Atom files now use spaces instead of tabs.

### CODE REVIEW

File: types/atom-keymap/README.md [hunk 1]
```diff
@@ -19,3 +19,19 @@ Many of the types used by Atom Keymap can be referenced from the AtomKeymap name
 ```ts
 function example(keybind: AtomKeymap.KeyBinding) {}
 ```
+
+### Exposing Private Methods and Properties
```
Comment (andy-ms) [inline types/atom-keymap/README.md]: We won't actually publish a custom README, maybe this should be moved to `index.d.ts`?

File: types/status-bar/README.md [hunk 2]
```diff
@@ -1 +1 @@
-***NOTE:*** *This package is deprecated and new versions have been merged into [`@types/atom`](https://www.npmjs.com/package/@types/atom). The most recent version can be accessed as part of the `Atom.Services` namespace.*
+***NOTE:*** *This package is deprecated and new versions have been merged into [`@types/atom`](https://www.npmjs.com/package/@types/atom). Type definitions for the most recent version can be accessed from within the "atom/status-bar" module.*
```
Comment (andy-ms) [inline types/status-bar/README.md]: Maybe this package should be deleted from DefinitelyTyped and deprecated?

File: types/text-buffer/tslint.json [hunk 3]
```diff
@@ -1,11 +1,12 @@
 {
     "extends": "dtslint/dt.json",
     "rules": {
-        // Custom rules.
         "class-name": true,
-        "indent": [true, "tabs"],
+        "indent": [true, "spaces", 4],
```
Comment (andy-ms) [inline types/text-buffer/tslint.json]: I wouldn't change existing formatting settings without a reason.

File: types/text-buffer/tslint.json [hunk 4]
```diff
@@ -1,11 +1,12 @@
...
         "jsdoc-format": true,
         "max-line-length": [true, 100],
+        "no-single-declare-module": false,
```
Comment (andy-ms) [inline types/text-buffer/tslint.json]: This rule shouldn't be ignored -- it's better to export things directly from a file than to use `declare module`.

File: types/atom-keymap/README.md (hunk 1 above)
Comment (GlenCFL) [reply types/atom-keymap/README.md]: This was modeled after the ```@types/angular``` definition. I'm aware that it wont be published to NPM, but there is a link from NPM directly to the definition page within the repository. I don't think it's the kind of thing that makes it directly into the definition file, especially not without markup.

I'll remove it if this isn't something you're interested in having within the repository.

File: types/status-bar/README.md (hunk 2 above)
Comment (GlenCFL) [reply types/status-bar/README.md]: It's a legacy definition that I have no part of really. It probably should be removed, but I don't really know how deprecation works here on DT with the types publisher. I can remove it with this commit if you'd like.

Comment (andy-ms) [reply types/status-bar/README.md]: It looks like the `status-bar` NPM package is something else entirely, so this should definitely be deprecated. #20727

File: types/text-buffer/tslint.json (hunk 3 above)
Comment (GlenCFL) [reply types/text-buffer/tslint.json]: I originally chose tabs for these definitions whenever I thought editorconfig files were fine within the repository. I'd prefer spaces now that those files are gone, but if I really have to convert back to tabs then I can.

There aren't exactly any other authors to care about the switch over.

Comment (andy-ms) [reply types/text-buffer/tslint.json]: OK, I was just worried that the next PR would change it back. If you're the only author then it's fine.

**Comment by andy-ms** [comment]
Thanks!

**Pull request merged**